xbmc/pvr/PVRChannelNumberInputHandler.cpp

```diff
@@ -44,54 +46,69 @@ void CPVRChannelNumberInputHandler::OnTimeout()
   OnInputDone();
 
   CSingleLock lock(m_mutex);
-  m_digits.clear();
-  m_strChannel.erase();
+  m_inputBuffer.erase();
 }
 
-void CPVRChannelNumberInputHandler::AppendChannelNumberDigit(int iDigit)
+void CPVRChannelNumberInputHandler::AppendChannelNumberCharacter(char cCharacter)
 {
-  if (iDigit < 0 || iDigit > 9)
+  if (cCharacter != CPVRChannelNumber::SEPARATOR && (cCharacter < '0' || cCharacter > '9'))
     return;
 
   CSingleLock lock(m_mutex);
 
-  if (m_digits.size() == (size_t)m_iMaxDigits)
-    m_digits.pop_front();
-
-  m_digits.emplace_back(iDigit);
-
-  // recalc channel string
-  m_strChannel.erase();
-  if (m_digits.size() != (size_t)m_iMaxDigits || GetChannelNumber() > 0)
+  if (cCharacter == CPVRChannelNumber::SEPARATOR)
   {
-    for (int digit : m_digits)
-      m_strChannel.append(StringUtils::Format("%d", digit));
+    // no leading separator
+    if (m_inputBuffer.empty())
+      return;
+
+    // max one separator
+    if (m_inputBuffer.find(CPVRChannelNumber::SEPARATOR) != std::string::npos)
+      return;
   }
 
+  if (m_inputBuffer.size() == static_cast<size_t>(m_iMaxDigits))
+    m_inputBuffer.erase(m_inputBuffer.begin());
+
+  m_inputBuffer.append(&cCharacter, 1);
+
   if (!m_timer.IsRunning())
     m_timer.Start(m_iDelay);
   else
     m_timer.Restart();
 }
 
-int CPVRChannelNumberInputHandler::GetChannelNumber() const
+CPVRChannelNumber CPVRChannelNumberInputHandler::GetChannelNumber() const
 {
-  int iNumber = 0;
-  int iDigitMultiplier = 1;
+  int iChannelNumber = 0;
+  int iSubChannelNumber = 0;
+
+  CSingleLock lock(m_mutex);
 
-  for (std::deque<int>::const_reverse_iterator it = m_digits.rbegin(); it != m_digits.rend(); ++it)
+  size_t pos = m_inputBuffer.find(CPVRChannelNumber::SEPARATOR);
+  if (pos != std::string::npos)
+  {
+    // main + sub
+    if (pos != 0)
+    {
+      iChannelNumber = std::atoi(m_inputBuffer.substr(0, pos).c_str());
+      if (pos != m_inputBuffer.back())
```

Comment:
@ksooo What is this comparison good for?
Comparing pos (size_t) with the last element of the string (char) doesn't seem valid.

Reply:
Ooops. I will fix this.